File: server/src/controllers/pageController.go
package controllers

import (
	"context"
	"net/http"
	"time"

	database "controll/src/db"
	helper "controll/src/helpers"
	"controll/src/models"

	"github.com/gin-gonic/gin"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
)

var pageCollection *mongo.Collection = database.OpenCollection(database.Client, "pages")

func CreatePage() gin.HandlerFunc {
	return func(c *gin.Context) {
		if ok, _, _ := helper.CheckAdminOrUidPermission(c, ""); !ok {
			return
		}

		var page models.Page
		if err := c.ShouldBindJSON(&page); err != nil {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Dados inválidos", "status": false})
			return
		}

		if page.Title == "" || page.Path == "" {
			c.JSON(http.StatusBadRequest, gin.H{"message": "Todos campos são obrigatórios", "status": false})
			return
		}

		var ctx, cancel = context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()

		count, err := pageCollection.CountDocuments(ctx, bson.M{"path": page.Path})
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao verificar dados", "status": false})
			return
		}
		if count > 0 {
			c.JSON(http.StatusConflict, gin.H{"message": "Página já cadastrada", "status": false})
			return
		}

		page.ID = primitive.NewObjectID()
		page.Author = c.MustGet("user").(models.User).ID
		_, err = pageCollection.InsertOne(ctx, page)
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao salvar a página", "status": false})
			return
		}

		c.JSON(http.StatusCreated, gin.H{"message": "Página criada com sucesso", "status": true, "page": page})
	}
}

func GetAllPages() gin.HandlerFunc {
	return func(c *gin.Context) {
		if ok, _, _ := helper.CheckAdminOrUidPermission(c, ""); !ok {
			return
		}

		var ctx, cancel = context.WithTimeout(context.Background(), 100*time.Second)
		defer cancel()

		cursor, err := pageCollection.Find(ctx, bson.M{})
		if err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao obter páginas", "status": false})
			return
		}
		defer cursor.Close(ctx)

		var pages []models.Page
		if err := cursor.All(ctx, &pages); err != nil {
			c.JSON(http.StatusInternalServerError, gin.H{"message": "Erro ao processar dados", "status": false})
			return
		}

		if len(pages) == 0 {
			c.JSON(http.StatusNotFound, gin.H{"message": "Sem páginas encontradas", "status": false})
			return
		}

		c.JSON(http.StatusOK, gin.H{"message": "Páginas encontradas com sucesso", "status": true, "pages": pages})
	}
}

File: server/src/Controllers/pageController.js
import Page from "../Models/pagesModel.js";

class PageController {
  static async createPage(req, res) {
    try {
      if (!req.user.user.isManager) {
        return res.status(403).json({
          message: "Sem permissão",
          status: false,
        });
      }

      const { title, path } = req.body;

      if (!title || !path) {
        return res.status(400).json({
          message: "Todos campos são obrigatórios",
          status: false,
        });
      }

      const isUnique = await Page.isUnique(path);

      if (!isUnique) {
        return res.status(409).json({
          message: "Página já cadastrada",
          status: false,
        });
      }

      const newPage = await Page.create({
        title,
        path,
        author: req.user.user._id,
      });

      if (!newPage) {
        return res.status(400).json({
          message: "Erro ao salvar a página",
          status: false,
        });
      }

      res.status(201).json({
        message: "Página criada com sucesso",
        status: true,
        page: newPage,
      });
    } catch (e) {
      res.status(500).json({
        message: "Erro interno do servidor",
        status: false,
      });
    }
  }
  static async getAll(req, res) {
    try {
      if (!req.user.user.isManager) {
        return res.status(403).json({
          message: "Sem permissão",
          status: false,
        });
      }

      const pages = await Page.find().exec();

      if (!pages) {
        return res.status(404).json({
          message: "Sem páginas encontradas",
          status: false,
        });
      }

      res.status(200).json({
        message: "Páginas encontradas com sucesso",
        status: true,
        pages,
      });
    } catch (e) {
      res.status(500).json({
        message: "Erro interno do servidor",
        status: false,
      });
    }
  }
}

export default PageController;
